{compiler}[system/system] Add CodePlay CUDA SYCL plug-in variants for oneAPI 2023.0.0 to 2025.0.0

Thyre opened the pull request:
CodePlay offers NVIDIA plug-ins for the Intel oneAPI compilers starting with oneAPI 2023.0.0.
Add separate EasyConfigs using these plug-ins by creating a bundle of the basic oneAPI compilers and the plug-in.
Since the plug-in needs to be installed inside the actual oneAPI installation, do not just use the oneAPI installation.

Requires:
- [x] https://github.com/easybuilders/easybuild-easyblocks/pull/3472.
- [x] https://github.com/easybuilders/easybuild-easyblocks/pull/3495 for oneAPI 2025.0.0
- [x] https://github.com/easybuilders/easybuild-easyblocks/pull/3509
    - [x] Requires https://github.com/easybuilders/easybuild-framework/pull/4707
- [x] https://github.com/easybuilders/easybuild-easyconfigs/pull/22206
- [x] https://github.com/easybuilders/easybuild-easyblocks/pull/3801

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 560.35.03, Python 3.12.7
See https://gist.github.com/Thyre/11de3dc3a2eb33afafe083a55372f43c for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
datenlager - Linux Ubuntu 24.04, x86_64, AMD Ryzen 7 3700X 8-Core Processor, Python 3.12.3
See https://gist.github.com/Thyre/aea1a28c0aadc46fb6719b8197a19390 for a full test report.

Thyre commented:
Test report by @Thyre
**SUCCESS**
Build succeeded for 8 out of 8 (1 easyconfigs in total)
datenlager - Linux Ubuntu 24.04, x86_64, AMD Ryzen 7 3700X 8-Core Processor, Python 3.12.3
See https://gist.github.com/Thyre/de8cb7aeeb85cb854e67db99a873b48b for a full test report.

SebastianAchilles commented:
Test report by @sebastianachilles
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
skl-rockylinux-810 - Linux Rocky Linux 8.10, x86_64, Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz (skylake), 1 x NVIDIA NVIDIA RTX A4000, 555.42.06, Python 3.6.8
See https://gist.github.com/SebastianAchilles/f392987fde98289c1f925ded8e39d3bb for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 560.35.03, Python 3.12.7
See https://gist.github.com/Thyre/74a1fc0bc6444b099ec82a9ddab94ce7 for a full test report.

SebastianAchilles commented:
@boegelbot please test @ jsc-zen3-a100
CORE_CNT=16
EB_ARGS='--include-easyblocks-from-pr 3472'

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
datenlager - Linux Ubuntu 24.04, x86_64, AMD Ryzen 7 3700X 8-Core Processor, Python 3.12.3
See https://gist.github.com/Thyre/dbb641d35fc23216e29143b5841fdf4d for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 0 out of 0 (1 easyconfigs in total)
jrlogin02.jureca - Linux Rocky Linux 9.4, x86_64, AMD EPYC 7742 64-Core Processor, 2 x NVIDIA Quadro RTX 8000, 560.35.03, Python 3.9.18
See https://gist.github.com/Thyre/d79081f410d1903b9231bc7a845f4f57 for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin02.jureca - Linux Rocky Linux 9.4, x86_64, AMD EPYC 7742 64-Core Processor, 2 x NVIDIA Quadro RTX 8000, 560.35.03, Python 3.9.18
See https://gist.github.com/Thyre/13624b3cbaa4390a914860f026ba6b49 for a full test report.

Thyre commented:
Added additional EasyConfigs for all `intel-compiler` versions we have starting with 2023.0.0.
Since 2023.1.0 needs GCC 12.3.0 and I haven't built that yet, testing will take just a bit longer.

The plug-in for oneAPI 2025.0.0 has seen some changes in the shared-library name and download path. Additionally, I do not see any headers in the extracted files anymore. Therefore, I removed it for this version.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472, https://github.com/easybuilders/easybuild-easyblocks/pull/3495
**SUCCESS**
Build succeeded for 5 out of 5 (5 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 565.57.01, Python 3.12.7
See https://gist.github.com/Thyre/0d371e210e63868625cdda8dae65f136 for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472, https://github.com/easybuilders/easybuild-easyblocks/pull/3495
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin02.jureca - Linux Rocky Linux 9.4, x86_64, AMD EPYC 7742 64-Core Processor, 2 x NVIDIA Quadro RTX 8000, 560.35.03, Python 3.9.18
See https://gist.github.com/Thyre/fb7f77092fdecdba796ea361714ca986 for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472, https://github.com/easybuilders/easybuild-easyblocks/pull/3495
**FAILED**
Build succeeded for 4 out of 6 (5 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 565.57.01, Python 3.12.7
See https://gist.github.com/Thyre/3f6b23df7ddb1bd701e11b37df413214 for a full test report.

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472, https://github.com/easybuilders/easybuild-easyblocks/pull/3495
**SUCCESS**
Build succeeded for 6 out of 6 (5 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 565.57.01, Python 3.12.7
See https://gist.github.com/Thyre/91d1fc94d6b2561c4a0d7259f4da15b3 for a full test report.

Thyre commented:
Here are some testing results with all the versions. I used the sample application described for oneAPI 2023.0.0 [here](https://developer.codeplay.com/products/oneapi/nvidia/2023.0.0/guides/get-started-guide-nvidia#run-a-sample-application):

**Code:**

<details>
<summary>Click to open</summary>

```cpp
#include <sycl/sycl.hpp>

int main() {
  // Creating buffer of 4 ints to be used inside the kernel code
  sycl::buffer<sycl::cl_int, 1> Buffer(4);

  // Creating SYCL queue
  sycl::queue Queue;

  // Size of index space for kernel
  sycl::range<1> NumOfWorkItems{Buffer.size()};

  // Submitting command group(work) to queue
  Queue.submit([&](sycl::handler &cgh) {
    // Getting write only access to the buffer on a device
    auto Accessor = Buffer.get_access<sycl::access::mode::write>(cgh);
    // Executing kernel
    cgh.parallel_for<class FillBuffer>(
        NumOfWorkItems, [=](sycl::id<1> WIid) {
          // Fill buffer with indexes
          Accessor[WIid] = (sycl::cl_int)WIid.get(0);
        });
  });

  // Getting read only access to the buffer on the host.
  // Implicit barrier waiting for queue to complete the work.
  const auto HostAccessor = Buffer.get_access<sycl::access::mode::read>();

  // Check the results
  bool MismatchFound = false;
  for (size_t I = 0; I < Buffer.size(); ++I) {
    if (HostAccessor[I] != I) {
      std::cout << "The result is incorrect for element: " << I
                << " , expected: " << I << " , got: " << HostAccessor[I]
                << std::endl;
      MismatchFound = true;
    }
  }

  if (!MismatchFound) {
    std::cout << "The results are correct!" << std::endl;
  }

  return MismatchFound;
}
```

</details>

---

**intel-compilers/2023.0.0-CUDA-11.7.0**

For oneAPI 2023.0.0, one may need to use the internal `clang++` instead of `icpx`, which is described [here](https://developer.codeplay.com/products/oneapi/nvidia/2023.0.0/guides/troubleshooting#bitcode-output-disabled-with-icpx).

```console
$  /data/EasyBuild-develop/software/intel-compilers/2023.0.0-CUDA-11.7.0/compiler/2023.0.0/linux/bin-llvm/clang++ -fsycl -fsycl-targets=nvptx64-nvidia-cuda simple-sycl-app.cpp -o simple-sycl-app
clang++: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
$ SYCL_DEVICE_FILTER=cuda SYCL_PI_TRACE=1 ./simple-sycl-app
SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_cuda.so [ PluginVersion: 11.15.1 ]
SYCL_PI_TRACE[all]: Selected device: -> final score = 1500
SYCL_PI_TRACE[all]:   platform: NVIDIA CUDA BACKEND
SYCL_PI_TRACE[all]:   device: NVIDIA GeForce RTX 3070
The results are correct!
```

**intel-compilers/2023.2.1-CUDA-12.2.2**

```console
$ icpx -fsycl -fsycl-targets=nvptx64-nvidia-cuda simple-sycl-app.cpp -o simple-sycl-app
[... several deprecation warnings ...]
$ SYCL_DEVICE_FILTER=cuda SYCL_PI_TRACE=1 ./simple-sycl-app                            

WARNING: The enviroment variable SYCL_DEVICE_FILTER is deprecated. Please use ONEAPI_DEVICE_SELECTOR instead.
For more details, please refer to:
https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#oneapi_device_selector

SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_cuda.so [ PluginVersion: 12.27.1 ]
SYCL_PI_TRACE[all]: Requested device_type: info::device_type::automatic
SYCL_PI_TRACE[all]: Selected device: -> final score = 1500
SYCL_PI_TRACE[all]:   platform: NVIDIA CUDA BACKEND
SYCL_PI_TRACE[all]:   device: NVIDIA GeForce RTX 3070
The results are correct!
```

**intel-compilers/2024.0.0-CUDA-12.3.0**

```console
$ icpx -fsycl -fsycl-targets=nvptx64-nvidia-cuda simple-sycl-app.cpp -o simple-sycl-app
[... several deprecation warnings ...]
$ SYCL_DEVICE_FILTER=cuda SYCL_PI_TRACE=1 ./simple-sycl-app                            

WARNING: The enviroment variable SYCL_DEVICE_FILTER is deprecated. Please use ONEAPI_DEVICE_SELECTOR instead.
For more details, please refer to:
https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#oneapi_device_selector

SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_cuda.so [ PluginVersion: 14.38.1 ]
SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_unified_runtime.so [ PluginVersion: 14.37.1 ]
SYCL_PI_TRACE[all]: Requested device_type: info::device_type::automatic
SYCL_PI_TRACE[all]: Selected device: -> final score = 1500
SYCL_PI_TRACE[all]:   platform: NVIDIA CUDA BACKEND
SYCL_PI_TRACE[all]:   device: NVIDIA GeForce RTX 3070
The results are correct!
```

**intel-compilers/2024.2.0-CUDA-12.5.0**
```console
$ icpx -fsycl -fsycl-targets=nvptx64-nvidia-cuda simple-sycl-app.cpp -o simple-sycl-app
[... several deprecation warnings ...]
$ SYCL_DEVICE_FILTER=cuda SYCL_PI_TRACE=1 ./simple-sycl-app

WARNING: The enviroment variable SYCL_DEVICE_FILTER is deprecated. Please use ONEAPI_DEVICE_SELECTOR instead.
For more details, please refer to:
https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#oneapi_device_selector

SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_cuda.so [ PluginVersion: 15.49.1 ]
SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_unified_runtime.so [ PluginVersion: 15.47.1 ]
SYCL_PI_TRACE[all]: Requested device_type: info::device_type::automatic
SYCL_PI_TRACE[all]: Selected device: -> final score = 1500
SYCL_PI_TRACE[all]:   platform: NVIDIA CUDA BACKEND
SYCL_PI_TRACE[all]:   device: NVIDIA GeForce RTX 3070
The results are correct!
```

**intel-compilers/2025.0.0-CUDA-12.6.0**
```console
$ icpx -fsycl -fsycl-targets=nvptx64-nvidia-cuda simple-sycl-app.cpp -o simple-sycl-app
[... several deprecation warnings ...]
$ ONEAPI_DEVICE_SELECTOR="cuda:*" SYCL_UR_TRACE=1 ./simple-sycl-app
<LOADER>[INFO]: loaded adapter 0x0xfe7ec00 (libur_adapter_cuda.so.0)
SYCL_UR_TRACE: Requested device_type: info::device_type::automatic
SYCL_UR_TRACE: Selected device: -> final score = 1500
SYCL_UR_TRACE:   platform: NVIDIA CUDA BACKEND
SYCL_UR_TRACE:   device: NVIDIA GeForce RTX 3070
The results are correct!
```

Thyre commented:
Test report by @Thyre
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/3472, https://github.com/easybuilders/easybuild-easyblocks/pull/3495
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jrlogin10.jureca - Linux Rocky Linux 9.4, x86_64, AMD EPYC 7742 64-Core Processor, 2 x NVIDIA Quadro RTX 8000, 560.35.03, Python 3.9.18
See https://gist.github.com/Thyre/6df91936a9367ff339a5cc4e6cb8547f for a full test report.

Thyre commented:
@boegelbot please test @ jsc-zen3-a100
CORE_CNT=16
EB_ARGS='--include-easyblocks-from-pr 3472'

Thyre commented:
Test report by @Thyre
**SUCCESS**
Build succeeded for 5 out of 5 (5 easyconfigs in total)
Linux - Linux EndeavourOS UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x NVIDIA NVIDIA GeForce RTX 3070, 565.57.01, Python 3.12.7
See https://gist.github.com/Thyre/7fa7ad9ec1a081ce1466930ec945a514 for a full test report.

Thyre commented:
@boegelbot please test @ jsc-zen3-a100
CORE_CNT=16

Thyre commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16
EB_BRANCH=5.0.x
EB_ARGS='--include-easyblocks-from-pr 3513'

Thyre commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16

Thyre commented:
Will add oneAPI 2025.1.0 variant once https://github.com/easybuilders/easybuild-easyconfigs/pull/22657 is merged.

Thyre commented:
> Will add oneAPI 2025.1.0 variant once #22657 is merged.

2025.1.0 proved a bit harder, since something in their scripts has changed. This PR contains a lot of EasyConfigs anyway, so merging them first, and looking into further versions later on, is probably a good idea..

Thyre commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16
EB_ARGS='--include-easyblocks-from-pr=3801 --installpath=/tmp/$USER/ebpr-21582'

Thyre commented:
Test report by @Thyre
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
Linux - Linux Arch Linux UNKNOWN, x86_64, AMD Ryzen 7 7800X3D 8-Core Processor, 1 x AMD Navi 48 [Radeon RX 9070/9070 XT/9070 GRE] (device id: 0x7550, gfx: gfx1201, driver: Linuxversion6.15.4-arch2-1(linux@archlinux)(gcc(GCC)15.1.120250425,GNUld(GNUBinutils)2.44.0)#1SMPPREEMPT_DYNAMICFri,27Jun202516:35:07+0000), 1 x AMD Raphael (device id: 0x164e, gfx: gfx1036, driver: Linuxversion6.15.4-arch2-1(linux@archlinux)(gcc(GCC)15.1.120250425,GNUld(GNUBinutils)2.44.0)#1SMPPREEMPT_DYNAMICFri,27Jun202516:35:07+0000), Python 3.13.5
See https://gist.github.com/Thyre/818da1c5d1dc3cf1e6502f0408f0da2f for a full test report.

Thyre commented:
@boegelbot please test @ jsc-zen3
CORE_CNT=16
EB_ARGS='--installpath=/tmp/$USER/ebpr-21582'
